lib/handler/mruby.c

```diff
@@ -720,85 +747,106 @@ void h2o_mruby_run_fiber(h2o_mruby_generator_t *generator, mrb_value receiver, m
         if (mrb->exc != NULL)
             goto GotException;
         status = mrb_fixnum(v);
+
+        /* if no special actions were necessary, then the output is a rack response */
+        if (status >= 0) break;
+
         /* take special action depending on the status code */
-        if (status < 0) {
-            if (status == H2O_MRUBY_CALLBACK_ID_EXCEPTION_RAISED) {
-                mrb->exc = mrb_obj_ptr(mrb_ary_entry(output, 1));
-                goto GotException;
+        if (status == H2O_MRUBY_CALLBACK_ID_EXCEPTION_RAISED) {
+            mrb->exc = mrb_obj_ptr(mrb_ary_entry(output, 1));
+            generator = h2o_mruby_get_generator(mrb, mrb_ary_entry(output, 2));
+            goto GotException;
+        } else if (status == H2O_MRUBY_CALLBACK_ID_CONFIGURING_APP) {
+            mrb_value pending = mrb_ary_new_capa(mrb, 2);
+            mrb_ary_set(mrb, pending, 0, receiver);
+            mrb_ary_set(mrb, pending, 1, input);
+            mrb_ary_push(mrb, ctx->pendings, pending);
```

Comment:
> Could you please elaborate how ctx->pendings is protected from GC?

In `on_context_init` , `ctx->pendings` is allocated before `mrb_gc_arena_save` is called. So arena index always indicates the succeeding position after it, thus I believe `ctx->pendings` is protected from gc (if I understand mruby's gc correctly)

> PS. Or, if you like, maybe we can have the pending list kept in the mruby-side. IIRC we have discussed the possibility before, but cannot recall why the idea was dismissed.

IIRC the conclusion was "Either's fine", but then I came to feel that handling this in c is more simple and clear than doing in mruby (to handle it in mruby, we have to find a dirty way to pass the generator argument to app proc keeping rack interface, otherwise we need to check whether `app` is initialized or not in `runner` fiber, which may decrease runtime  performance, etc..)